examples/microphone_nextjs/src/pages/index.tsx

```diff
@@ -0,0 +1,282 @@
+import { GetServerSideProps } from "next";
+import React, {
+  useState,
+  useMemo,
+  useRef,
+  CSSProperties,
+  useEffect,
+} from "react";
+import { RealtimeSession, RealtimeRecognitionResult } from "speechmatics";
+import { AudioRecorder, useAudioDevices } from "../utils/recorder";
+
+type MainProps = { jwt?: string; simplUI?: boolean };
+
+type SessionState = "configure" | "starting" | "blocked" | "error" | "running";
+
+export default function Main({ jwt }: MainProps) {
+  const [transcription, setTranscription] = useState<
+    RealtimeRecognitionResult[]
+  >([]);
+  const [audioDeviceId, setAudioDeviceId] = useState<string>("");
+  const [sessionState, setSessionState] = useState<SessionState>("configure");
+
+  const rtSessionRef = useRef(new RealtimeSession(jwt));
+
+  // Get devices using our custom hook
+  const [requestDevices, setRequestDevices] = useState<boolean>(false);
+  const [devices, denied] = useAudioDevices(requestDevices);
+
+  // useEffect listens for changes in devices
+  // It sets a default deviceId if no valid deviceId is already set
+  useEffect(() => {
+    if (
+      devices.length &&
+      !devices.some((item) => item.deviceId == audioDeviceId)
+    )
+      setAudioDeviceId(devices[0].deviceId);
+    if (denied) setSessionState("blocked");
+  }, [devices, denied]);
+
+  // sendAudio is used as a wrapper for the websocket to check the socket is finished init-ing before sending data
+  const sendAudio = (data: Float32Array | Buffer) => {
+    if (
+      rtSessionRef.current.rtSocketHandler &&
+      rtSessionRef.current.isConnected
+    ) {
+      rtSessionRef.current.sendAudio(data);
+    }
+  };
+
+  // Memoise AudioRecorder so it doesn't get recreated on re-render
+  const audioRecorder = useMemo(() => new AudioRecorder(sendAudio), []);
+
+  // Attach our event listeners to the realtime session
+  rtSessionRef.current.addListener("AddTranscript", (res) => {
+    setTranscription([...transcription, ...res.results]);
+  });
+
+  // start audio recording once the websocket is connected
+  rtSessionRef.current.addListener("RecognitionStarted", async () => {
+    setSessionState("running");
+  });
+
+  rtSessionRef.current.addListener("EndOfTranscript", async () => {
+    setSessionState("configure");
+    await audioRecorder.stopRecording();
+  });
+
+  rtSessionRef.current.addListener("Error", async () => {
+    setSessionState("error");
+    await audioRecorder.stopRecording();
+  });
+
+  // Call the start method on click to start the websocket
+  const startTranscription = async () => {
+    setSessionState("starting");
+    await audioRecorder.startRecording(audioDeviceId)
+    .then(async () => {
```

Comment:
Do we need a `.then` here?

Reply:
I put that in so that we can capture the scenario specifically when start recording fails (not just when the entire start up process fails)